presto-thrift/src/test/java/io/prestosql/plugin/thrift/integration/TestThriftProjectionPushdown.java

```diff
@@ -0,0 +1,188 @@
+import com.google.common.collect.ImmutableSet;
+import io.airlift.drift.server.DriftServer;
+import io.prestosql.Session;
+import io.prestosql.connector.CatalogName;
+import io.prestosql.metadata.TableHandle;
+import io.prestosql.plugin.thrift.ThriftColumnHandle;
+import io.prestosql.plugin.thrift.ThriftPlugin;
+import io.prestosql.plugin.thrift.ThriftTableHandle;
+import io.prestosql.plugin.thrift.ThriftTransactionHandle;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.connector.ConnectorTableHandle;
+import io.prestosql.spi.connector.SchemaTableName;
+import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.sql.parser.SqlParser;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.TypeAnalyzer;
+import io.prestosql.sql.planner.iterative.rule.PushProjectionIntoTableScan;
+import io.prestosql.sql.planner.iterative.rule.test.BaseRuleTest;
+import io.prestosql.sql.planner.plan.Assignments;
+import io.prestosql.sql.tree.SymbolReference;
+import io.prestosql.testing.LocalQueryRunner;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.Test;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.prestosql.plugin.thrift.integration.ThriftQueryRunner.driftServerPort;
+import static io.prestosql.plugin.thrift.integration.ThriftQueryRunner.startThriftServers;
+import static io.prestosql.spi.type.VarcharType.VARCHAR;
+import static io.prestosql.sql.planner.assertions.ObjectMatcher.equalityMatcher;
+import static io.prestosql.sql.planner.assertions.PlanMatchPattern.expression;
+import static io.prestosql.sql.planner.assertions.PlanMatchPattern.project;
+import static io.prestosql.sql.planner.assertions.PlanMatchPattern.tableScanWithMatchers;
+import static io.prestosql.testing.TestingSession.testSessionBuilder;
+import static java.util.stream.Collectors.joining;
+
+public class TestThriftProjectionPushdown
+        extends BaseRuleTest
+{
+    private static final String THRIFT_CATALOG = "thrift";
+    private static final String TINY_SCHEMA = "tiny";
+    private List<DriftServer> servers;
+
+    private static final ThriftTableHandle NATION_THRIFT_HANDLE = new ThriftTableHandle(new SchemaTableName(TINY_SCHEMA, "nation"));
+
+    private static final TableHandle NATION_TABLE_HANDLE = new TableHandle(
+            new CatalogName(THRIFT_CATALOG),
+            NATION_THRIFT_HANDLE,
+            ThriftTransactionHandle.INSTANCE,
+            Optional.empty());
+
+    private static final Session THRIFT_SESSION = testSessionBuilder().setCatalog(THRIFT_CATALOG).setSchema(TINY_SCHEMA).build();
+
+    @Override
+    protected Optional<LocalQueryRunner> createLocalQueryRunner()
+    {
+        try {
+            servers = startThriftServers(1, false);
+        }
+        catch (Throwable t) {
+            for (DriftServer server : servers) {
```

Comment:
if an exception is thrown `servers` will be set to `null` here. Also you should catch exceptions here and add them as suppressed.

Reply:
good catch.